### PR TITLE
Rust optimized state

### PR DESCRIPTION
**Pull request by petertdavies**
This PR removes the old python optimized state and replaces it with the new rust optimized state.

The new optimized state consists of three components:
* The python code in `src/ethereum_optimized`.
* The [`ethereum-pyspec-db`](https://github.com/petertdavies/rust-pyspec-db) rust package.
* The [`rust-pyspec-glue`](https://github.com/petertdavies/rust-pyspec-glue), which is a python package that contains a mix of python and rust code.

`rust-pyspec-glue` is uploaded to PyPI and will be installed automatically with `pip install ethereum[optimized]`. I have not got binary wheels working yet so currently it will be built from source which requires a rust toolchain and clang.

Everything should work as before, but the database format has completely changed so any old databases you have should be deleted.

I have also added an `--optimized` option to `pytest` that runs the tests with optimized state and ethash (this is slightly faster, but mostly due to ethash).

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: src/ethereum_optimized/byzantium/state_db.py [hunk 1]
```diff
@@ -14,47 +14,23 @@
...
-from .trie_utils import decode_to_internal_node, encode_internal_node_nohash
-
-try:
-    import lmdb
-except ImportError as e:
-    # Add a message, but keep it an ImportError.
-    raise e from Exception(
-        "Install with `pip install 'ethereum[optimized]'` to enable this "
-        "package"
-    )
```
Comment (SamWilsn) [inline src/ethereum_optimized/byzantium/state_db.py:51]: Should we keep this `try`/`except` around `import rust_pyspec_glue`?

File: src/ethereum_optimized/byzantium/state_db.py [hunk 2]
```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
...
-        raise Exception("Current transaction is not a db transaction")
-    state._tx_stack.pop().abort()
-    state._dirty_accounts = [{}]
-    state._dirty_storage = [{}]
-    state._destroyed_accounts = [set()]
-    if state._tx_stack != []:
-        state._current_tx = state._tx_stack[-1]
-    else:
-        state._current_tx = None
+    if len(state.tx_restore_points) != 0:
```
Comment (SamWilsn) [inline src/ethereum_optimized/byzantium/state_db.py:122]:
```suggestion
    if state.tx_restore_points:
```

File: src/ethereum_optimized/byzantium/state_db.py [hunk 3]
```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
...
-                # Higher levels refer to the account prior to destruction
-                return U256(0)
-    res = state._current_tx.get(b"\x01" + address + b"\x00" + key)
-    if res is None:
-        return U256(0)
-    else:
-        res = rlp.decode(res)
-        assert isinstance(res, bytes)
-        return U256.from_be_bytes(res)
+    if len(state.tx_restore_points) != 0:
```
Comment (SamWilsn) [inline src/ethereum_optimized/byzantium/state_db.py:152]:
```suggestion
    if state.tx_restore_points:
```

File: src/ethereum_optimized/byzantium/state_db.py [hunk 4]
```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
...
+    for address in state.destroyed_accounts:
+        state.db.destroy_storage(address)
+    for address, account in state.dirty_accounts.items():
+        state.db.set_account(address, account)
+    for address, storage in state.dirty_storage.items():
+        for key, value in storage.items():
+            state.db.set_storage(address, key, value)
+    state.destroyed_accounts = set()
+    state.dirty_accounts = {}
+    state.dirty_storage = {}
```
Comment (SamWilsn) [inline src/ethereum_optimized/byzantium/state_db.py:163]: Would using `.clear()` be better for these? I think it causes less (de)allocations.

File: src/ethereum_optimized/byzantium/state_db.py [hunk 5]
```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
...
     See `ethereum.byzantium.state`.
     """
-    if account is None:
-        state._dirty_accounts[-1][address] = None
-    else:
-        state._dirty_accounts[-1][address] = account
+    restore_point = state.tx_restore_points.pop()
+    while len(state.journal) > restore_point:
+        item = state.journal.pop()
+        if len(item) == 3:
```
Comment (SamWilsn) [inline src/ethereum_optimized/byzantium/state_db.py:204]: Hm... I'd like to either see proper types for each of these branches `isinstance(item, SomeDescriptiveType)` or comments describing what each branch represents.

**Review by SamWilsn — APPROVED** [review]

**Comment by petertdavies** [comment]
@SamWilsn None of the changes you've requested to rust-pyspec-db will affect this PR, assuming you are happy for the function names in rust-pyspec-glue to stay the same.

**Pull request merged**